File: ios/runanywhere-sdks-latest/examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/ui/screens/models/ModelDisplay.kt
package com.runanywhere.runanywhereai.ui.screens.models

import ai.runanywhere.proto.v1.InferenceFramework
import ai.runanywhere.proto.v1.ModelCategory
import androidx.compose.ui.graphics.vector.ImageVector
import com.runanywhere.sdk.npu.qhexrt.QHexRT
import com.runanywhere.runanywhereai.ui.theme.icons.Brand
import com.runanywhere.runanywhereai.ui.theme.icons.RACBrands
import com.runanywhere.runanywhereai.ui.theme.icons.RACIcons
import com.runanywhere.sdk.public.extensions.Models.displayName
import com.runanywhere.sdk.public.types.RAModelInfo

// Pure model -> display mappers. No Compose, no state.

private val privateHfTags = setOf("private", "requires-hf-auth", "hf-auth", "gated")

enum class ConsumerModelGroup(val title: String, val footer: String) {
    CHAT_MODELS(
        "Chat Models",
        "Primary assistants for private chat. Download one to use it offline.",
    ),
    APPLE_BUILT_IN(
        "Built-in Models",
        "Models provided by the device or operating system need no download.",
    ),
    VISION_MODELS(
        "Vision Models",
        "Photo, camera, and multimodal understanding models.",
    ),
    VOICE_MODELS(
        "Voice Models",
        "Speech, dictation, and read-aloud models.",
    ),
    DOCUMENT_MODELS(
        "Document Models",
        "Embedding and answer models used by document Q&A.",
    ),
    LORA_ADAPTERS(
        "LoRA Adapters",
        "Style, task, and domain adapters that attach to compatible chat models.",
    ),
    OTHER(
        "Other Models",
        "Additional SDK entries available on this device.",
    ),
}

fun RAModelInfo.brand(): Brand {
    if (framework == InferenceFramework.INFERENCE_FRAMEWORK_FOUNDATION_MODELS ||
        framework == InferenceFramework.INFERENCE_FRAMEWORK_SYSTEM_TTS
    ) {
        return RACBrands.Foundation
    }
    val n = name.lowercase()
    return when {
        "llama" in n -> RACBrands.Meta
        "mistral" in n -> RACBrands.Mistral
        "qwen" in n -> RACBrands.Qwen
        "lfm" in n || "liquid" in n -> RACBrands.Liquid
        "whisper" in n -> RACBrands.Whisper
        else -> RACBrands.HuggingFace
    }
}

fun InferenceFramework.shortLabel(): String = when (this) {
    InferenceFramework.INFERENCE_FRAMEWORK_LLAMA_CPP -> "Llama CPP"
    InferenceFramework.INFERENCE_FRAMEWORK_ONNX -> "ONNX"
    InferenceFramework.INFERENCE_FRAMEWORK_FOUNDATION_MODELS -> "Apple"
    InferenceFramework.INFERENCE_FRAMEWORK_SYSTEM_TTS -> "System"
    InferenceFramework.INFERENCE_FRAMEWORK_QHEXRT -> "NPU"
    InferenceFramework.INFERENCE_FRAMEWORK_SHERPA -> "Sherpa"
    InferenceFramework.INFERENCE_FRAMEWORK_COREML -> "Core ML"
    InferenceFramework.INFERENCE_FRAMEWORK_MLX -> "MLX"
    InferenceFramework.INFERENCE_FRAMEWORK_PIPER_TTS -> "Piper"
    InferenceFramework.INFERENCE_FRAMEWORK_FLUID_AUDIO -> "Fluid"
    InferenceFramework.INFERENCE_FRAMEWORK_TFLITE -> "TFLite"
    InferenceFramework.INFERENCE_FRAMEWORK_EXECUTORCH -> "ExecuTorch"
    InferenceFramework.INFERENCE_FRAMEWORK_MEDIAPIPE -> "MediaPipe"
    InferenceFramework.INFERENCE_FRAMEWORK_MLC -> "MLC"
    InferenceFramework.INFERENCE_FRAMEWORK_PICO_LLM -> "Pico"
    InferenceFramework.INFERENCE_FRAMEWORK_SWIFT_TRANSFORMERS -> "Swift"
    InferenceFramework.INFERENCE_FRAMEWORK_BUILT_IN -> "Built-in"
    InferenceFramework.INFERENCE_FRAMEWORK_NONE -> "None"
    InferenceFramework.INFERENCE_FRAMEWORK_UNKNOWN -> "Unknown"
    else -> displayName
}

fun InferenceFramework.consumerBackendLabel(): String = when (this) {
    InferenceFramework.INFERENCE_FRAMEWORK_LLAMA_CPP -> "Llama CPP"
    InferenceFramework.INFERENCE_FRAMEWORK_ONNX -> "ONNX Runtime"
    InferenceFramework.INFERENCE_FRAMEWORK_FOUNDATION_MODELS -> "Apple Built-in"
    InferenceFramework.INFERENCE_FRAMEWORK_SYSTEM_TTS -> "System Voice"
    InferenceFramework.INFERENCE_FRAMEWORK_QHEXRT -> "Hexagon NPU"
    InferenceFramework.INFERENCE_FRAMEWORK_SHERPA -> "Sherpa Voice"
    InferenceFramework.INFERENCE_FRAMEWORK_COREML -> "Core ML"
    InferenceFramework.INFERENCE_FRAMEWORK_MLX -> "MLX"
    InferenceFramework.INFERENCE_FRAMEWORK_PIPER_TTS -> "Piper Voice"
    InferenceFramework.INFERENCE_FRAMEWORK_FLUID_AUDIO -> "Fluid Audio"
    InferenceFramework.INFERENCE_FRAMEWORK_TFLITE -> "TensorFlow Lite"
    InferenceFramework.INFERENCE_FRAMEWORK_EXECUTORCH -> "ExecuTorch"
    InferenceFramework.INFERENCE_FRAMEWORK_MEDIAPIPE -> "MediaPipe"
    InferenceFramework.INFERENCE_FRAMEWORK_MLC -> "MLC"
    InferenceFramework.INFERENCE_FRAMEWORK_PICO_LLM -> "Pico LLM"
    InferenceFramework.INFERENCE_FRAMEWORK_SWIFT_TRANSFORMERS -> "Swift Transformers"
    InferenceFramework.INFERENCE_FRAMEWORK_BUILT_IN -> "Built-in"
    InferenceFramework.INFERENCE_FRAMEWORK_NONE -> "No Backend"
    InferenceFramework.INFERENCE_FRAMEWORK_UNKNOWN -> "Unknown Backend"
    else -> displayName
}

fun InferenceFramework.backendIcon(): ImageVector = when (this) {
    InferenceFramework.INFERENCE_FRAMEWORK_LLAMA_CPP -> RACIcons.Outline.Stack
    InferenceFramework.INFERENCE_FRAMEWORK_ONNX -> RACIcons.Outline.Activity
    InferenceFramework.INFERENCE_FRAMEWORK_FOUNDATION_MODELS -> RACIcons.Filled.Bolt
    InferenceFramework.INFERENCE_FRAMEWORK_SYSTEM_TTS -> RACIcons.Outline.Robot
    InferenceFramework.INFERENCE_FRAMEWORK_QHEXRT -> RACIcons.Outline.Cpu
    InferenceFramework.INFERENCE_FRAMEWORK_SHERPA -> RACIcons.Outline.Microphone
    InferenceFramework.INFERENCE_FRAMEWORK_COREML -> RACIcons.Outline.Cpu
    InferenceFramework.INFERENCE_FRAMEWORK_MLX -> RACIcons.Outline.Cpu
    InferenceFramework.INFERENCE_FRAMEWORK_PIPER_TTS -> RACIcons.Outline.Robot
    InferenceFramework.INFERENCE_FRAMEWORK_FLUID_AUDIO -> RACIcons.Outline.Activity
    InferenceFramework.INFERENCE_FRAMEWORK_TFLITE,
    InferenceFramework.INFERENCE_FRAMEWORK_EXECUTORCH,
    InferenceFramework.INFERENCE_FRAMEWORK_MEDIAPIPE,
    InferenceFramework.INFERENCE_FRAMEWORK_MLC,
    InferenceFramework.INFERENCE_FRAMEWORK_PICO_LLM,
    InferenceFramework.INFERENCE_FRAMEWORK_SWIFT_TRANSFORMERS,
    -> RACIcons.Outline.Stack
    InferenceFramework.INFERENCE_FRAMEWORK_BUILT_IN -> RACIcons.Outline.Check
    InferenceFramework.INFERENCE_FRAMEWORK_NONE -> RACIcons.Outline.Close
    InferenceFramework.INFERENCE_FRAMEWORK_UNKNOWN -> RACIcons.Outline.InfoCircle
    else -> RACIcons.Outline.Stack
}

fun RAModelInfo.consumerGroup(): ConsumerModelGroup = when {
    framework == InferenceFramework.INFERENCE_FRAMEWORK_FOUNDATION_MODELS ||
        framework == InferenceFramework.INFERENCE_FRAMEWORK_SYSTEM_TTS -> ConsumerModelGroup.APPLE_BUILT_IN
    isLoraAdapterEntry() -> ConsumerModelGroup.LORA_ADAPTERS
    category == ModelCategory.MODEL_CATEGORY_LANGUAGE -> ConsumerModelGroup.CHAT_MODELS
    category == ModelCategory.MODEL_CATEGORY_MULTIMODAL ||
        category == ModelCategory.MODEL_CATEGORY_VISION -> ConsumerModelGroup.VISION_MODELS
    category == ModelCategory.MODEL_CATEGORY_SPEECH_RECOGNITION ||
        category == ModelCategory.MODEL_CATEGORY_SPEECH_SYNTHESIS ||
        category == ModelCategory.MODEL_CATEGORY_AUDIO ||
        category == ModelCategory.MODEL_CATEGORY_VOICE_ACTIVITY_DETECTION -> ConsumerModelGroup.VOICE_MODELS
    category == ModelCategory.MODEL_CATEGORY_EMBEDDING -> ConsumerModelGroup.DOCUMENT_MODELS
    else -> ConsumerModelGroup.OTHER
}

private fun RAModelInfo.isLoraAdapterEntry(): Boolean =
    id.startsWith("lora-adapter:") ||
        metadata?.tags.orEmpty().any { tag ->
            tag.equals("lora-adapter", ignoreCase = true) ||
                tag.equals("lora", ignoreCase = true)
        }

// A consumer-facing "family" — a group of related model variants (e.g. Qwen3, LFM2,
// Whisper). The picker shows one card per family; tapping expands its variants.
data class ModelFamily(val key: String, val title: String, val tagline: String)

// Ordered family matchers. First match wins; each entry maps id/name substrings to a
// friendly family. Kept as a pure list so families are easy to extend without touching
// UI. HNPU (on-device accelerated) families are distinguished from their GGUF cousins
// by the "_HNPU"/"(HNPU)" marker so the two never collapse into one card.
private data class FamilyMatcher(
    val key: String,
    val title: String,
    val tagline: String,
    val matches: (String) -> Boolean,
)

private fun containsAll(vararg needles: String): (String) -> Boolean =
    { h -> needles.all { it in h } }

private fun containsAny(vararg needles: String): (String) -> Boolean =
    { h -> needles.any { it in h } }

private val familyMatchers: List<FamilyMatcher> = listOf(
    // On-device accelerated (HNPU) families first — the marker keeps them separate.
    FamilyMatcher("qwen3-npu", "Qwen3 (On-device)", "Accelerated Qwen chat", containsAll("qwen3", "hnpu")),
    FamilyMatcher("lfm2-npu", "LFM2 (On-device)", "Accelerated Liquid chat", containsAll("lfm2", "hnpu")),
    FamilyMatcher("llama-npu", "Llama (On-device)", "Accelerated Meta chat", containsAll("llama", "hnpu")),
    FamilyMatcher("gemma-npu", "Gemma (On-device)", "Accelerated Google chat", containsAll("gemma", "hnpu")),
    FamilyMatcher("deepseek-npu", "DeepSeek (On-device)", "Accelerated reasoning", containsAll("deepseek", "hnpu")),
    FamilyMatcher("nemotron-npu", "Nemotron (On-device)", "Accelerated NVIDIA models", containsAll("nemotron", "hnpu")),
    FamilyMatcher("qwenvl-npu", "Qwen-VL (On-device)", "Accelerated vision chat", containsAll("qwen3_vl", "hnpu")),
    FamilyMatcher("whisper-npu", "Whisper (On-device)", "Accelerated speech-to-text", containsAll("whisper", "hnpu")),
    FamilyMatcher("kokoro-npu", "Kokoro (On-device)", "Accelerated read-aloud", containsAll("kokoro", "hnpu")),
    FamilyMatcher("kitten-npu", "Kitten (On-device)", "Accelerated read-aloud", containsAll("kitten", "hnpu")),
    FamilyMatcher("embed-npu", "On-device Embeddings", "Accelerated document search", containsAll("embed", "hnpu")),

    // Standard on-device families.
    FamilyMatcher("qwen3", "Qwen3", "Fast, capable chat", containsAny("qwen3", "qwen 3")),
    FamilyMatcher("qwen25", "Qwen2.5", "Reliable everyday chat", containsAny("qwen2.5", "qwen 2.5")),
    FamilyMatcher("qwen2vl", "Qwen2-VL", "Chat about photos", containsAny("qwen2-vl", "qwen2.5-vl")),
    FamilyMatcher("lfm2", "LFM2", "Efficient Liquid chat", containsAny("lfm2")),
    FamilyMatcher("smol", "SmolLM / SmolVLM", "Tiny, quick assistants", containsAny("smol")),
    FamilyMatcher("llama", "Llama", "Meta's open chat", containsAny("llama")),
    FamilyMatcher("mistral", "Mistral", "Balanced open chat", containsAny("mistral")),
    FamilyMatcher("gemma", "Gemma", "Google's open chat", containsAny("gemma")),
    FamilyMatcher("whisper", "Whisper", "Speech-to-text", containsAny("whisper")),
    FamilyMatcher("piper", "Piper", "Natural read-aloud", containsAny("piper")),
    FamilyMatcher("minilm", "MiniLM", "Document search", containsAny("minilm")),
    FamilyMatcher("silero", "Silero", "Detects speech", containsAny("silero")),
)

// Pure family derivation from id + name. Falls back to a category-based family so every
// model lands somewhere sensible.
fun RAModelInfo.family(): ModelFamily {
    val haystack = "$id $name".lowercase()
    familyMatchers.firstOrNull { it.matches(haystack) }?.let {
        return ModelFamily(it.key, it.title, it.tagline)
    }
    return when (consumerGroup()) {
        ConsumerModelGroup.VISION_MODELS -> ModelFamily("other-vision", "Vision Models", "Chat about photos")
        ConsumerModelGroup.VOICE_MODELS -> ModelFamily("other-voice", "Voice Models", "Speech and read-aloud")
        ConsumerModelGroup.DOCUMENT_MODELS -> ModelFamily("other-docs", "Document Models", "Search your documents")
        ConsumerModelGroup.LORA_ADAPTERS -> ModelFamily("other-lora", "Adapters", "Fine-tune a chat model")
        ConsumerModelGroup.CHAT_MODELS,
        ConsumerModelGroup.APPLE_BUILT_IN -> ModelFamily("other-chat", "Other Chat Models", "More assistants")
        ConsumerModelGroup.OTHER -> ModelFamily("other", "Other Models", "Additional models")
    }
}


fun RAModelInfo.requiresHfAuth(): Boolean {
    val tags = metadata?.tags.orEmpty().map { it.lowercase() }
    return (framework == InferenceFramework.INFERENCE_FRAMEWORK_QHEXRT &&
        QHexRT.modelRequiresHfAuth(id)) || tags.any { it in privateHfTags }
}

// Effective on-disk / in-memory footprint used for sizing and consumer tags.
fun RAModelInfo.effectiveBytes(): Long = when {
    download_size_bytes > 0L -> download_size_bytes
    (memory_required_bytes ?: 0L) > 0L -> memory_required_bytes ?: 0L
    else -> 0L
}

// Rough "intelligence" hint derived from parameter count (parsed from the name)
// with a byte-size fallback. Used INTERNALLY only (recommendation + variant order);
// never shown as a raw label on cards.
enum class ModelIntelligence { LITE, BALANCED, SMART, GENIUS }

fun RAModelInfo.intelligence(): ModelIntelligence {
    val params = estimatedParamsBillions()
    if (params != null) {
        return when {
            params < 0.7 -> ModelIntelligence.LITE
            params < 2.0 -> ModelIntelligence.BALANCED
            params < 5.0 -> ModelIntelligence.SMART
            else -> ModelIntelligence.GENIUS
        }
    }
    // No parseable param count — fall back to footprint.
    val mb = effectiveBytes() / 1_048_576.0
    return when {
        mb < 500 -> ModelIntelligence.LITE
        mb < 2_000 -> ModelIntelligence.BALANCED
        mb < 6_000 -> ModelIntelligence.SMART
        else -> ModelIntelligence.GENIUS
    }
}

// Parses a parameter count in billions from tokens like "0.6b", "1.5b", "350m", "7b".
private fun RAModelInfo.estimatedParamsBillions(): Double? {
    val haystack = "$id $name".lowercase()
    Regex("""(\d+(?:\.\d+)?)\s*b\b""").find(haystack)?.let {
        return it.groupValues[1].toDoubleOrNull()
    }
    Regex("""(\d+(?:\.\d+)?)\s*m\b""").find(haystack)?.let {
        return it.groupValues[1].toDoubleOrNull()?.div(1000.0)
    }
    return null
}

// The single "feel" word shown on every card. Derived from intelligence but collapsed
// to three consumer-friendly buckets — no Lite/Genius, no raw parameter counts.
enum class ModelFeel(val label: String) { FAST("Fast"), BALANCED("Balanced"), SMART("Smart") }

fun RAModelInfo.feel(): ModelFeel = when (intelligence()) {
    ModelIntelligence.LITE -> ModelFeel.FAST
    ModelIntelligence.BALANCED -> ModelFeel.BALANCED
    ModelIntelligence.SMART, ModelIntelligence.GENIUS -> ModelFeel.SMART
}

// Semantic kind so the UI can color pills consistently across the picker.
enum class ConsumerTagKind { FEEL, CAPABILITY }

data class ConsumerTag(val label: String, val kind: ConsumerTagKind)

// AT MOST two clean, consumer-facing tags per card: a feel tag, plus one notable
// capability tag when applicable. No quantization, size, context length, or backend
// terms ever appear here — those stay internal to recommendation / variant ordering.
fun RAModelInfo.consumerTags(): List<ConsumerTag> = buildList {
    add(ConsumerTag(feel().label, ConsumerTagKind.FEEL))
    notableCapability()?.let { add(ConsumerTag(it, ConsumerTagKind.CAPABILITY)) }
}

// The single most notable capability, or null when nothing stands out. Capability
// (vision/voice/documents) wins over behavioral hints (tools/thinking).
private fun RAModelInfo.notableCapability(): String? {
    when (category) {
        ModelCategory.MODEL_CATEGORY_MULTIMODAL,
        ModelCategory.MODEL_CATEGORY_VISION,
        -> return "Vision"
        ModelCategory.MODEL_CATEGORY_SPEECH_RECOGNITION,
        ModelCategory.MODEL_CATEGORY_SPEECH_SYNTHESIS,
        ModelCategory.MODEL_CATEGORY_AUDIO,
        -> return "Voice"
        ModelCategory.MODEL_CATEGORY_EMBEDDING -> return "Documents"
        else -> Unit
    }
    if (isToolCallingModel()) return "Great for tools"
    if (supports_thinking) return "Thinks"
    return null
}

// Tool-oriented chat models: LiquidAI "tool" builds and NVIDIA Nemotron families
// are trained/tuned for function calling; also catch any id/name flagged "tool".
private fun RAModelInfo.isToolCallingModel(): Boolean {
    val haystack = "$id $name".lowercase()
    return "tool" in haystack || "nemotron" in haystack
}

// Friendly variant label used inside a family's variant list — "Smaller · faster" vs
// "Larger · smarter" — instead of quant strings. Ordered by footprint upstream.
fun RAModelInfo.variantFeelLabel(): String = when (feel()) {
    ModelFeel.FAST -> "Smaller · faster"
    ModelFeel.BALANCED -> "Balanced"
    ModelFeel.SMART -> "Larger · smarter"
}

fun formatModelSize(bytes: Long): String {
    if (bytes <= 0) return "—"
    val gb = bytes / 1_073_741_824.0
    return if (gb >= 1.0) "%.2f GB".format(gb) else "%.0f MB".format(bytes / 1_048_576.0)
}

// Quant / precision tokens stripped from display titles. Word-bounded so parameter
// counts like "0.5B" or family names like "LFM2" are never touched.
private val quantTokenRegex = Regex("""(?i)\b(Q\d(_K(_[MS])?|_0)?|FP16|F16|[458]BIT|DWQ)\b""")

// Vendor / marker noise removed from display titles; the brand icon and backend badge
// carry that information instead. "Instruct" is implied for consumer chat models.
private val titleNoise = listOf("(HNPU)", "(ONNX)", "LiquidAI", "Instruct")

// Clean, consumer-facing row title: family + parameter size (e.g. "LFM2 1.2B Tool",
// "Qwen3 4B") with quant suffixes and vendor prefixes stripped from the raw name.
fun RAModelInfo.displayTitle(): String {
    var title = name
    titleNoise.forEach { title = title.replace(it, "", ignoreCase = true) }
    title = quantTokenRegex.replace(title, "")
    return title.replace(Regex("""\s+"""), " ").trim().trimEnd('-', '·', ',').trim()
}

// Download-size label shown on every model row — the user always sees the cost.
fun RAModelInfo.sizeLabel(): String {
    val bytes = effectiveBytes()
    if (bytes > 0) return formatModelSize(bytes)
    return if (requiresHfAuth()) "Size varies" else "Size unknown"
}

File: ios/runanywhere-sdks-latest/examples/web/RunAnywhereAI/src/services/model-display.ts
/**
 * Shared display helpers for model catalog rendering.
 *
 * Previously these label tables were inlined in both `components/model-selection.ts`
 * and `views/storage.ts`, and a third byte formatter lived in `components/dialogs.ts`.
 * Keeping a single canonical site avoids the picker, the storage tab, and the
 * eviction dialog drifting from each other when proto enums add new values.
 */

import { ModelCategory } from '@runanywhere/web';
import type { CatalogEntry } from './model-catalog';
export { formatFramework } from '@runanywhere/web';

/**
 * Returns the HTML entity for the emoji shown next to a model row. The
 * return value is an HTML-safe entity ("&#129302;") so it can be inlined
 * inside an innerHTML template without further escaping.
 */
export function modalityEmoji(category: ModelCategory): string {
  switch (category) {
    case ModelCategory.MODEL_CATEGORY_LANGUAGE:
      return '&#129302;';
    case ModelCategory.MODEL_CATEGORY_MULTIMODAL:
      return '&#128065;';
    case ModelCategory.MODEL_CATEGORY_SPEECH_RECOGNITION:
      return '&#127908;';
    case ModelCategory.MODEL_CATEGORY_SPEECH_SYNTHESIS:
      return '&#128266;';
    case ModelCategory.MODEL_CATEGORY_VOICE_ACTIVITY_DETECTION:
      return '&#128483;';
    case ModelCategory.MODEL_CATEGORY_IMAGE_GENERATION:
      return '&#127912;';
    case ModelCategory.MODEL_CATEGORY_EMBEDDING:
      return '&#128279;';
    default:
      return '&#9881;&#65039;';
  }
}

/**
 * Decimal byte formatter ("GB" / "MB" / "KB"). Aligns with how model
 * catalogs advertise file sizes (1 GB = 10^9 bytes) and with the eviction
 * dialog's storage gauge, both of which run against model catalog byte inputs.
 */
export function formatBytes(bytes: number): string {
  if (bytes >= 1_000_000_000) return `${(bytes / 1_000_000_000).toFixed(1)} GB`;
  if (bytes >= 1_000_000) return `${Math.round(bytes / 1_000_000)} MB`;
  return `${Math.round(bytes / 1_000)} KB`;
}

export function modelDisplaySizeBytes(model: {
  downloadSizeBytes?: number;
  memoryRequiredBytes?: number;
}): number {
  return model.downloadSizeBytes && model.downloadSizeBytes > 0
    ? model.downloadSizeBytes
    : model.memoryRequiredBytes ?? 0;
}

const BACKEND_FORMAT_TOKENS = new Set(['(ONNX)', '(GGUF)', '(MLX)']);

/**
 * Consumer-facing model name with quantization/technical suffixes stripped
 * (e.g. "SmolLM2 360M Q8_0" → "SmolLM2 360M", "Qwen3 4B Q4_K_M" → "Qwen3 4B").
 * Pure string cleanup — never changes which model an id refers to.
 */
export function cleanModelName(name: string): string {
  return name
    .split(/\s+/)
    .filter((token) => !BACKEND_FORMAT_TOKENS.has(token.toUpperCase()))
    .join(' ')
    // Quant tokens: Q8_0, Q4_K_M, Q6_K, F16, BF16, DWQ, 4bit/8-bit, int8…
    .replace(/\b(?:Q\d+(?:_[A-Z0-9]+)*|BF16|F16|F32|DWQ|INT[48]|\d+\s?-?bits?)\b/gi, '')
    // Collapse leftover separators/whitespace from the removal.
    .replace(/\s{2,}/g, ' ')
    .replace(/[\s\-·(]+$/g, '')
    .trim();
}

// ---------------------------------------------------------------------------
// Consumer tags — minimal, plain-language pills. Deliberately hides all
// technical detail (quantization, context length, inference backend). A card
// shows AT MOST two pills: one intelligence/size "feel" and, only when notable,
// one capability. Kept pure so any view shares the same vocabulary.
// ---------------------------------------------------------------------------

/**
 * The visual family a tag belongs to. The picker maps each kind to a distinct
 * pill color so users can scan "how it feels" vs. "what it can do" at a glance.
 */
export type ConsumerTagKind = 'feel' | 'capability';

export interface ConsumerTag {
  label: string;
  kind: ConsumerTagKind;
}

/** Intelligence/size feel — exactly one per model. */
export type ModelFeel = 'Fast' | 'Balanced' | 'Smart';

/** Notable capability — at most one per model, only when it stands out. */
export type ModelCapability = 'Great for tools' | 'Thinks' | 'Vision' | 'Voice' | 'Documents';

const GB = 1_000_000_000;

/**
 * At most two consumer tags: the feel first, then one capability when notable.
 * Never surfaces size class, quant, or backend text.
 */
export function consumerTags(entry: CatalogEntry): ConsumerTag[] {
  const tags: ConsumerTag[] = [{ label: modelFeel(entry), kind: 'feel' }];
  const capability = modelCapability(entry);
  if (capability) tags.push({ label: capability, kind: 'capability' });
  return tags;
}

/**
 * Intelligence/size feel from the parameter count in the name (e.g. "0.6B",
 * "360M"), falling back to advertised bytes. <0.7B → Fast, <2B → Balanced,
 * otherwise Smart.
 */
export function modelFeel(entry: CatalogEntry): ModelFeel {
  const params = parseParamBillions(entry.name);
  const billions = params ?? bytesToApproxParams(modelDisplaySizeBytes(entry));
  if (billions < 0.7) return 'Fast';
  if (billions < 2) return 'Balanced';
  return 'Smart';
}

/**
 * The single most notable capability, or `null` when nothing stands out.
 * Tool-calling and thinking win over modality tags because they are the
 * differentiators a consumer cares about when picking a chat model.
 */
export function modelCapability(entry: CatalogEntry): ModelCapability | null {
  const haystack = `${entry.id} ${entry.name}`.toLowerCase();
  if (haystack.includes('tool')) return 'Great for tools';
  if (entry.supportsThinking) return 'Thinks';
  return categoryCapability(entry.category);
}

function categoryCapability(category: ModelCategory): ModelCapability | null {
  switch (category) {
    case ModelCategory.MODEL_CATEGORY_MULTIMODAL:
      return 'Vision';
    case ModelCategory.MODEL_CATEGORY_SPEECH_RECOGNITION:
    case ModelCategory.MODEL_CATEGORY_SPEECH_SYNTHESIS:
      return 'Voice';
    case ModelCategory.MODEL_CATEGORY_EMBEDDING:
      return 'Documents';
    default:
      return null;
  }
}

/** Extract a parameter count in billions from strings like "4B" / "360M". */
function parseParamBillions(name: string): number | null {
  const match = name.match(/(\d+(?:\.\d+)?)\s*([bm])\b/i);
  if (!match) return null;
  const value = Number.parseFloat(match[1]);
  if (!Number.isFinite(value)) return null;
  return match[2].toLowerCase() === 'b' ? value : value / 1000;
}

/** Very rough params-from-bytes estimate for entries lacking a name hint. */
function bytesToApproxParams(bytes: number): number {
  return bytes / GB;
}

// ---------------------------------------------------------------------------
// Model families — group catalog entries into consumer-facing product lines
// (e.g. "Qwen3", "LFM2", "Whisper") so the picker can show one card per family
// and reveal individual variants on tap. Pure, name/id-driven.
// ---------------------------------------------------------------------------

export interface ModelFamily {
  /** Stable key derived from the model id/name (e.g. "qwen3"). */
  key: string;
  /** Consumer-facing display name (e.g. "Qwen3"). */
  name: string;
  /** Friendly one-liner describing what the family is good for. */
  tagline: string;
}

/**
 * Ordered family matchers. Each tests the lowercased "id + name" haystack; the
 * first match wins. Order matters — more specific ids (qwen3, qwen2.5) precede
 * generic ones. A trailing catch-all guarantees every entry lands somewhere.
 */
const FAMILY_MATCHERS: ReadonlyArray<{
  key: string;
  name: string;
  tagline: string;
  test: RegExp;
}> = [
  { key: 'qwen3', name: 'Qwen3', tagline: 'Latest Qwen chat models with a thinking mode.', test: /qwen3/ },
  { key: 'qwen2.5', name: 'Qwen2.5', tagline: 'Compact, capable all-round chat models.', test: /qwen2\.?5/ },
  { key: 'qwen2-vl', name: 'Qwen2-VL', tagline: 'Qwen models that can also see images.', test: /qwen2-?vl/ },
  { key: 'llama', name: 'Llama', tagline: "Meta's versatile open chat models.", test: /llama/ },
  { key: 'lfm2-vl', name: 'LFM2-VL', tagline: 'LiquidAI models that read images and text.', test: /lfm2-?vl/ },
  { key: 'lfm2', name: 'LFM2', tagline: 'LiquidAI models tuned for fast on-device chat.', test: /lfm2/ },
  { key: 'smolvlm', name: 'SmolVLM', tagline: 'Tiny vision-language models for quick demos.', test: /smolvlm/ },
  { key: 'smollm', name: 'SmolLM', tagline: 'Very small, speedy instruction models.', test: /smollm/ },
  { key: 'whisper', name: 'Whisper', tagline: 'Turns speech into text, on device.', test: /whisper/ },
  { key: 'piper', name: 'Piper', tagline: 'Natural-sounding text-to-speech voices.', test: /piper|vits/ },
  { key: 'silero', name: 'Silero', tagline: 'Detects when someone is speaking.', test: /silero|vad/ },
  { key: 'minilm', name: 'MiniLM', tagline: 'Powers document search and memory.', test: /minilm|embedding/ },
];

const FALLBACK_FAMILY: ModelFamily = {
  key: 'other',
  name: 'Other models',
  tagline: 'Additional on-device models.',
};

/** Derive the consumer-facing family for a catalog entry. Never throws. */
export function modelFamily(entry: CatalogEntry): ModelFamily {
  const haystack = `${entry.id} ${entry.name}`.toLowerCase();
  const match = FAMILY_MATCHERS.find((family) => family.test.test(haystack));
  if (!match) return FALLBACK_FAMILY;
  return { key: match.key, name: match.name, tagline: match.tagline };
}

/**
 * A friendly, quant-free size feel for a single variant, used when a family is
 * expanded so the user compares options by experience, not by quant string.
 */
export function variantSizeFeel(entry: CatalogEntry): string {
  const bytes = modelDisplaySizeBytes(entry);
  if (bytes < 0.35 * GB) return 'Smallest · fastest';
  if (bytes < 0.7 * GB) return 'Smaller · faster';
  if (bytes < 1.5 * GB) return 'Balanced';
  return 'Larger · smarter';
}
